development/webpack/utils/helpers.ts

```diff
@@ -0,0 +1,243 @@
+ */
+export const noop = () => undefined;
+
+/**
+ *
+ * @returns Returns the current version of MetaMask as specified in package.json
+ */
+export const getMetaMaskVersion = (): string => {
+  const { version } = require('../../../package.json');
+  return version as string;
+};
+
+/**
+ * Collects all entry files for use with webpack.
+ *
+ * TODO: move this logic into the ManifestPlugin
+ *
+ * @param manifest - Base manifest file
+ * @param appRoot - Absolute directory to search for entry files listed in the
+ * base manifest
+ * @returns an `entry` object containing html and JS entry points for use with
+ * webpack, and an array, `manifestScripts`, list of filepaths of all scripts
+ * that were added to it.
+ */
+export function collectEntries(manifest: Manifest, appRoot: string) {
+  const entry: EntryObject = {};
+  /**
+   * Scripts that must be self-contained and not split into chunks.
+   */
+  const selfContainedScripts: Set<string> = new Set([
+    // Snow shouldn't be chunked
+    'snow.prod',
+    'use-snow',
+  ]);
+
+  function addManifestScript(filename?: string) {
+    if (filename) {
+      selfContainedScripts.add(filename);
+      entry[filename] = {
+        chunkLoading: false,
+        filename, // output filename
+        import: join(appRoot, filename), // the path to the file to use as an entry
+      };
+    }
+  }
+
+  function addHtml(filename?: string) {
+    if (filename) {
+      assertValidEntryFileName(filename, appRoot);
+      entry[parse(filename).name] = join(appRoot, filename);
+    }
+  }
+
+  // add content_scripts to entries
+  manifest.content_scripts?.forEach((s) => s.js?.forEach(addManifestScript));
+
+  if (manifest.manifest_version === 3) {
+    addManifestScript(manifest.background?.service_worker);
+    manifest.web_accessible_resources?.forEach(({ resources }) =>
+      resources.forEach((filename) => {
+        filename.endsWith('.js') && addManifestScript(filename);
+      }),
+    );
+  } else {
+    manifest.web_accessible_resources?.forEach((filename) => {
+      filename.endsWith('.js') && addManifestScript(filename);
+    });
+    manifest.background?.scripts?.forEach(addManifestScript);
+    addHtml(manifest.background?.page);
+  }
+
+  for (const filename of readdirSync(appRoot)) {
+    // ignore non-htm/html files
+    if (/\.html?$/iu.test(filename)) {
+      addHtml(filename);
+    }
+  }
+
+  /**
+   * Ignore scripts that were found in the manifest, as these are only loaded by
+   * the browser extension platform itself.
+   *
+   * @param chunk
+   * @param chunk.name
+   * @returns
+   */
+  function canBeChunked({ name }: Chunk): boolean {
+    return !name || !selfContainedScripts.has(name);
+  }
+  return { entry, canBeChunked };
+}
+
+/**
+ * @param filename
+ * @param appRoot
+ * @throws Throws an `Error` if the file is an invalid entrypoint filename
+ * (a file starting with "_")
+ */
+function assertValidEntryFileName(filename: string, appRoot: string) {
```

Comment:
This build tool doesn't do much validation at build time, but entrypoint filenames at the root that start with `_` bit me during development, so I added in the validation here.

If this check wasn't here, and someone created a root entry point file that started with `_` chrome would error when adding the package, so we could remove this check and rely on that if anyone thinks this validation isn't necessary.